File: Middle_School_Management_App/Middle_School_Management_App/ScheduleController.swift
//
//  ScheduleController.swift
//  Middle_School_Management_App
//
//  Created by Vivianne Yee on 2023-11-11.
//

import Foundation
import RealmSwift

class ScheduleController {

    init() {}
    
    enum NetworkError: Error {
        case emptyResponse
        case invalidResponse
        case serverError(String)
    }
    
    // create a new schedule
    func createSchedule(day1: [String], day2: [String], day3: [String], day4: [String], day5: [String], day6: [String], day7: [String], day8: [String], day9: [String], day10: [String], completion: @escaping (Result<Data, Error>) -> Void) {
        let url = URL(string: "http://localhost:3000/schedule/schedules")!
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.addValue("application/json", forHTTPHeaderField: "Content-Type")

                
        let body: [String: Any] = [
            "day1": day1,
            "day2": day2,
            "day3": day3,
            "day4": day4,
            "day5": day5,
            "day6": day6,
            "day7": day7,
            "day8": day8,
            "day9": day9,
            "day10": day10
        ]
        
        do {
            let jsonData = try JSONSerialization.data(withJSONObject: body)
            request.httpBody = jsonData
        } catch {
            completion(.failure(error))
            return
        }
        
        URLSession.shared.dataTask(with: request) { data, response, error in
            // handle error
            if let error = error {
                completion(.failure(error))
                print("Could not create schedule")
                return
            }
            
            // check for successful response status
            guard let httpResponse = response as? HTTPURLResponse, httpResponse.statusCode == 201 else {
                completion(.failure(NetworkError.invalidResponse))
                return
            }
            
            // check that response data is not empty
            guard let responseData = data else {
                completion(.failure(NetworkError.emptyResponse))
                return
            }
            
            completion(.success(responseData))
        }.resume()
    }
    
    // get a schedule by its id
    func getScheduleById(id: String, completion: @escaping (Result<Schedule, Error>) -> Void) {
        let url = URL(string: "http://localhost:3000/schedule/schedules/\(id)")!
        var request = URLRequest(url: url)
        request.httpMethod = "GET"
        
        URLSession.shared.dataTask(with: request) { data, response, error in
            // handle error
            if let error = error {
                completion(.failure(error))
                print("Could not find schedule")
                return
            }
            
            // check for successful response status
            guard let httpResponse = response as? HTTPURLResponse, httpResponse.statusCode == 200 else {
                completion(.failure(NetworkError.invalidResponse))
                return
            }
            
            // check that response data is not empty
            guard let responseData = data else {
                completion(.failure(NetworkError.emptyResponse))
                return
            }
            
            do {
                let decoder = JSONDecoder()
                let scheduleObject = try decoder.decode(Schedule.self, from: responseData)
                completion(.success(scheduleObject))
            } catch {
                completion(.failure(error))
            }
                        
        }.resume()
    }
    
    // update a schedule
    func updateSchedule(id: String, day1: [String], day2: [String], day3: [String], day4: [String], day5: [String], day6: [String], day7: [String], day8: [String], day9: [String], day10: [String], completion: @escaping (Result<Schedule, Error>) -> Void) {
        let url = URL(string: "http://localhost:3000/schedule/schedules/\(id)")!
        var request = URLRequest(url: url)
        request.httpMethod = "PUT"
        request.addValue("application/json", forHTTPHeaderField: "Content-Type")
        
        let body: [String: Any] = [
            "day1": day1,
            "day2": day2,
            "day3": day3,
            "day4": day4,
            "day5": day5,
            "day6": day6,
            "day7": day7,
            "day8": day8,
            "day9": day9,
            "day10": day10
        ]
        
        do {
            let jsonData = try JSONSerialization.data(withJSONObject: body)
            request.httpBody = jsonData
        } catch {
            completion(.failure(error))
            return
        }
        
        URLSession.shared.dataTask(with: request) { data, response, error in
            // handle error
            if let error = error {
                completion(.failure(error))
                print("Could not update schedule")
                return
            }
            
            // check for successful response status
            guard let httpResponse = response as? HTTPURLResponse, httpResponse.statusCode == 200 else {
                completion(.failure(NetworkError.invalidResponse))
                return
            }
            
            // check that response data is not empty
            guard let responseData = data else {
                completion(.failure(NetworkError.emptyResponse))
                return
            }
            
            do {
                let decoder = JSONDecoder()
                let scheduleObject = try decoder.decode(Schedule.self, from: responseData)
                completion(.success(scheduleObject))
            } catch {
                completion(.failure(error))
            }
            
        }.resume()
    }
    
    // delete a schedule
    func deleteSchedule(id: String, completion: @escaping (Result<Data, Error>) -> Void) {
        let url = URL(string: "http://localhost:3000/schedule/schedules/\(id)")!
        var request = URLRequest(url: url)
        request.httpMethod = "DELETE"
        
        URLSession.shared.dataTask(with: request) { data, response, error in
            // handle error
            if let error = error {
                completion(.failure(error))
                print("Could not delete schedule")
                return
            }
            
            // check for successful response status
            guard let httpResponse = response as? HTTPURLResponse, httpResponse.statusCode == 200 else {
                completion(.failure(NetworkError.invalidResponse))
                return
            }
            
            // check that response data is not empty
            guard let responseData = data else {
                completion(.failure(NetworkError.emptyResponse))
                return
            }
            
            completion(.success(responseData))
            
        }.resume()
    }
}

File: Backend/controllers/scheduleController.js
const Schedule = require('../models/Schedule')

// create a new schedule
exports.createSchedule = async (req, res) => {

    const { day1, day2, day3, day4, day5, day6, day7, day8, day9, day10 } = req.body

    try {
        // create new schedule
        const newSchedule = new Schedule({
            day1: day1,
            day2: day2,
            day3: day3,
            day4: day4,
            day5: day5,
            day6: day6,
            day7: day7,
            day8: day8,
            day9: day9,
            day10: day10
        })

        await newSchedule.save()

        return res.status(201).json({ message: 'Schedule created successfully' })
    } catch (error) {
        // catch server error
        console.error(error)
        return res.status(500).json({ message: 'Internal Server Error' })
    }
}

// get schedule by id
exports.getScheduleById = async (req, res) => {
    const scheduleId = req.params.id

    try {
        const retrievedSchedule = await Schedule.findById(scheduleId)

        if (!retrievedSchedule) {
            return res.status(404),json({ error: 'Schedule not found' })
        } else {
            return res.status(200).json({ message: 'Schedule retrieved successfully', 
                _id: retrievedSchedule._id,
                day1: retrievedSchedule.day1,
                day2: retrievedSchedule.day2,
                day3: retrievedSchedule.day3,
                day4: retrievedSchedule.day4,
                day5: retrievedSchedule.day5,
                day6: retrievedSchedule.day6,
                day7: retrievedSchedule.day7,
                day8: retrievedSchedule.day8,
                day9: retrievedSchedule.day9,
                day10: retrievedSchedule.day10
            })
        }
    } catch (error) {
        // catch server error
        console.error(error)
        return res.status(500).json({ error: 'Internal Server Error' })
    }
}

// update schedule
exports.updateSchedule = async (req, res) => {
    const scheduleId = req.params.id
    const { day1, day2, day3, day4, day5, day6, day7, day8, day9, day10 } = req.body

    try {
        // find the schedule by id and update schedules for each day that are changed otherwise update with original data
        const updateSchedule = await Schedule.findByIdAndUpdate(scheduleId, { day1, day2, day3, day4, day5, day6, day7, day8, day9, day10 }, { new: true })

        if (!updateSchedule) {
            // 404 if schedule with given id does not exist
            return res.status(404).json({ error: 'Schedule not found' })
        }

        return res.status(200).json({ message: 'Schedule updated successfully', 
            _id: retrievedSchedule._id,
            day1: retrievedSchedule.day1,
            day2: retrievedSchedule.day2,
            day3: retrievedSchedule.day3,
            day4: retrievedSchedule.day4,
            day5: retrievedSchedule.day5,
            day6: retrievedSchedule.day6,
            day7: retrievedSchedule.day7,
            day8: retrievedSchedule.day8,
            day9: retrievedSchedule.day9,
            day10: retrievedSchedule.day10
        })
    } catch (error) {
        // catch server error 
        console.error(error)
        res.status(500).json({ error: 'Internal Server Error' })
    }
}

// delete schedule
exports.deleteSchedule = async (req, res) => {
    const scheduleId = req.params.id

    try {
        // find the schedule by id and delete
        const deletedSchedule = await Schedule.findByIdAndDelete(scheduleId)

        if (!deletedSchedule) {
            // 404 if schedule with given id does not exist
            return res.status(404).json({ error: 'Schedule not found' })
        }

        res.status(200).json({ message: 'Schedule deleted successfully' })
    } catch (error) {
        // catch server error
        console.error(error)
        res.status(500).json({ error: 'Internal Server Error' })
    }
}
